### PR TITLE
store: make RawTrieNode borsh serialisable

### PR DESCRIPTION
**Pull request by mina86**
Replace Children::encode_into and decode methods by implementations of
BorshSerializable and BorshDeserializable traits and with that change
RawTrieNode into an enum for which borsh implementations can be
derived.  While the first step is quite straightforward, the second
requires RawTrieNode::Branch variant being split into two:
BranchNoValue and BranchWithValue.

The main benefit of this is less code (the commit has delta of -68
lines) and API consistent with other serialised objects.  However,
this doesn’t come without downsides.

Firstly, having separate BranchNoValue and BranchWithValue branches
means that it’s now a bit more awkward to handle branch as a single
case with common code paths having to be refactored.

Secondly, one might imagine in the future a new representation for
nodes which embed value.  For example {4u8}{key_len:u32}{key}{value}
might be added for a leaf.  This would essentially mean that value
comprises of the rest of encoded bytes.  The issue for borsh is that
it would like to know length of {value} before it reads it.  Either
redundant {value_len:u32} would need to be included or some hacks
employed (namely not using borsh for RawTrieNodeWithSize and then
consuming the rest of the reader when decoding RawTrieNode).

Lastly, I’m not certain about performance implications.  Borsh might
actually be slower than hand-implemented decoder which operates on
a slice.  If this is actually the case I haven’t measured.

### CODE REVIEW

File: core/store/src/trie/raw_node.rs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    children[3] = Some(super::Trie::EMPTY_ROOT);
+    let node = RawTrieNode::BranchNoValue(children);
+    let encoded = [
+        1, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+        0, 0, 0, 0, 0, 42, 0, 0, 0, 0, 0, 0, 0,
+    ];
+    test(node, &encoded);
+
+    let mut children = Children::default();
+    children[3] = Some(super::Trie::EMPTY_ROOT);
```
Comment (robin-near) [inline core/store/src/trie/raw_node.rs:147]: No issue with your code, but this test seemed a bit lacking - I think we should at least test for (1) multiple children, (2) first children set, (3) last children set

**Review by robin-near — APPROVED** [review]
Looks much cleaner, thanks!

File: core/store/src/trie/raw_node.rs (hunk 1 above)
Comment (mina86) [reply core/store/src/trie/raw_node.rs]: Done.

**Comment by mina86** [comment]
Can I haz S-automerge?

**Comment by mina86** [comment]
Friendly ping.

**Pull request merged**